zabbix/resource_zabbix_item.go

```diff
@@ -0,0 +1,263 @@
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  0,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					v := val.(int)
+					if v < 0 || v > 4 {
+						errs = append(errs, fmt.Errorf("%q, must be between 0 and 4 inclusive, got %d", key, v))
+					}
+					return
+				},
+			},
+			"data_type": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  0,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					v := val.(int)
+					if v < 0 || v > 3 {
+						errs = append(errs, fmt.Errorf("%q, must be between 0 and 3 inclusive, got %d", key, v))
+					}
+					return
+				},
+			},
+			"delta": &schema.Schema{
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Value that will be stored. ",
+				Default:     0,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					v := val.(int)
+					if v < 0 || v > 2 {
+						errs = append(errs, fmt.Errorf("%q, must be between 0 and 2 inclusive, got %d", key, v))
+					}
+					return
+				},
+			},
+			"description": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of the item. ",
+				Default:     "",
+			},
+			"history": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Number of days to keep item's history data. Default: 90. ",
+				Default:     "90",
+			},
+			"trends": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Number of days to keep item's trends data. Default: 365. ",
+				Default:     "365",
+			},
+			"trapper_host": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Allowed hosts. Used only by trapper items. ",
+			},
+		},
+	}
+}
+
+func createItemObject(d *schema.ResourceData, api *zabbix.API) (*zabbix.Item, error) {
+
+	item := zabbix.Item{
+		ItemID:       d.Get("item_id").(string),
+		Delay:        d.Get("delay").(int),
+		HostID:       d.Get("host_id").(string),
+		InterfaceID:  d.Get("interface_id").(string),
+		Key:          d.Get("key").(string),
+		Name:         d.Get("name").(string),
+		Type:         zabbix.ItemType(d.Get("type").(int)),
+		ValueType:    zabbix.ValueType(d.Get("value_type").(int)),
+		DataType:     zabbix.DataType(d.Get("data_type").(int)),
+		Delta:        zabbix.DeltaType(d.Get("delta").(int)),
+		Description:  d.Get("description").(string),
+		History:      d.Get("history").(string),
+		Trends:       d.Get("trends").(string),
+		TrapperHosts: d.Get("trapper_host").(string),
+	}
+	return &item, nil
+}
+
+func resourceZabbixItemCreate(d *schema.ResourceData, meta interface{}) error {
+	api := meta.(*zabbix.API)
+
+	item, err := createItemObject(d, api)
+	if err != nil {
+		return err
+	}
+	items := zabbix.Items{*item}
+
+	return resource.Retry(time.Minute, func() *resource.RetryError {
```

Comment:
This whole retry logic should be extracted in a function reused for all zabbix API requests.